prov/shm/src/smr_rma.c

```diff
@@ -204,7 +207,7 @@ ssize_t smr_generic_rma(struct smr_ep *ep, const struct iovec *iov,
 							     pend, resp);
 				}
 			} else if (total_len <= smr_env.sar_threshold ||
-			    iface != FI_HMEM_SYSTEM) {
+			    iface != FI_HMEM_SYSTEM || op_flags & FI_INJECT) {
```

Comment:
This seems to be using SAR protocol for inject. I wonder if that is correct? should not we always use `smr_format_inject` for inject? I also have a question regarding earlier code (around line 160).

```
if (total_len <= SMR_INJECT_SIZE &&
		   !(op_flags & FI_DELIVERY_COMPLETE) && !use_ipc)
```
It seems if FI_DELIVERY_COMPLETE is requested, we do not use  `smr_format_inject`. I wonder why is that?

Reply:
Delivery complete requires a completion when the message arrives - the smr_format_inject protocol assumes FI_INJECT_COMPLETE completion semantics so if the user requests a higher completion level, we can't use that protocol. The SAR protocol is basically using an inject bounce buffer (like the inject protocol) but also has an accompanying response entry to support the delivery complete semantics. As long as the message length is less than the inject size, then the SAR protocol will behave pretty much exactly the same as the inject protocol + response entry

Reply:
I guess I do not understand FI_INJECT_COMPLETE.

From my understanding of the API, when FI_INJECT is specified, libfabric will not write a completion entry. Meanwhile, FI_DELIVERY_COMPLETE is about the timing of the completion entry libfabric wrote. So if libfabric does not write a completion entry (inject), it should not matter.

Reply:
FI_INJECT and FI_INJECT_COMPLETE are different flags.  FI_INJECT indicates that the buffer must be re-usable upon return from the call.  It does NOT impact the completion of the operation.

Reply:
The inject API calls do not generate completions. If FI_INJECT is specified it just means that the application can reuse the buffer immediately after the send call returns. The completion levels refer to at what point a completion is generated. So you could still set FI_INJECT so that you can reuse the buffer immediately but also set FI_DELIVERY_COMPLETE so that you get the completion later. Or, if FI_INJECT and FI_INJECT_COMPLETE were set, then the buffer could be immediately reused and a completion entry would immediately be generated.
From the man pages which might be helpful:
`Conceptually, this means that the fi_inject function behaves as if the FI_INJECT transfer flag were set, selective completions are enabled, and the FI_COMPLETION flag is not specified.`

Reply:
Ah. I see. Thanks for the clarification!